lms/djangoapps/verify_student/views.py

```diff
@@ -1050,6 +1051,14 @@ def _submit_attempt(self, user, face_image, photo_id_image=None, initial_verific
         attempt.mark_ready()
         attempt.submit(copy_id_photo_from=initial_verification)
 
+        idv_update_signal.send(
```

Comment:
I am wondering, what happen if the submit attempt is not successful? I feel like at this place, the status of the attempt will not be `submitted`, rather, it would be `ready` rather than `submitted`. The signal should be fired from the model rather than the view?

Reply:
So there are a couple different approaches to this: 
1. send a signal for each function that updates a status for the SoftwareSecurePhotoVerification model, like you're suggesting
2. Override the save method for the model. If we override the method, we can check to see which fields have been changed. If the status field has changed, send a signal. 
3. Have a pre_save signal receiver for the SoftwareSecurePhotoVerification defined in the name affirmation code. Within that receiver we can detect if a field is going to be changed and decide what to do from there. I've read that this isn't necessarily a recommended approach because it is pre_save, meaning that if something were to go wrong while saving the model, we may have already fired off some logic based on the pre save information.

Option 1 is more closely aligned to what's being done in the edx-proctoring PR, but option 2 or 3 would probably be cleaner.

Reply:
But yeah, looking specifically at submitting it definitely makes sense for that signal to be fired from the model, not from the view. I can update the code to align with the first approach I mentioned.

Reply:
2 is exactly why the old exam attempt history was terrible, it tried to be smart about what changes would be interesting. So I like 1.

Certainly in this case if it might send submitted even on a failure that's not OK.

Reply:
if we're tying directly to the model why not use post_save?

Reply:
I think in order to tell what fields were updated if we use a post_save receiver, we would need to look at `update_fields` which would be in the kwargs passed to the receiver - more about this here: https://stackoverflow.com/questions/36719566/identify-the-changed-fields-in-django-post-save-signal/55005137

In order for `update_fields` to not be `None`, you would also need to modify what's being passed to Model.save() - more on this here: https://stackoverflow.com/questions/54578488/django-signals-kwargsupdate-fields-is-always-none-on-model-update-via-djan

Because I don't think the sender and instance differ in a post_save receiver, I hadn't really considered it as an option.

Reply:
Alternatively, we could just have a post_save receiver and not care about what fields have changed.

Reply:
Also if we are tying IDV directly to the model, potentially with a post_save receiver, should we use the same approach for proctored exam attempts?

Reply:
My belief is asking "which fields have been updated" is always a horrible dark hole full of snakes and if we can avoid it we should. If we put it in post-save we're declaring we're not going to be smart, we're just emitting signals on save. If we want to be smart I think your original approach of having it only in a few higher level methods would be correct.

Reply:
I think it would be fine to just not look into the changed fields on post_save and notify name affirmation about any change. If we're pushing all the responsibility to the name affirmation app it should be resilient to getting a status that doesn't always reflect a change anyway.